src/intersections/boxes.jl

```diff
@@ -34,3 +34,46 @@ function intersecttype(f::Function, b1::Box{Dim,T}, b2::Box{Dim,T}) where {Dim,T
     return FaceTouchingBoxes(Box(u, v)) |> f
   end
 end
+
+"""
+    intersecttype(function, ray, box)
+
+Calculate the intersection type of a `ray`` and a `box`` and apply function `f` to it.
+
+## References
+
+* Williams A, Barrus S, Morley R K, et al., 2005. [An efficient and robust ray-box
+  intersection algorithm]
+  (https://dl.acm.org/doi/abs/10.1145/1198555.1198748)
+"""
+function intersecttype(f::Function, r::Ray{Dim,T}, b::Box{Dim,T}) where {Dim,T}
+  invdir = one(T) ./ direction(r)
+  lo, up = coordinates.(extrema(b))
+  orig = coordinates(origin(r))
+
+  tmin = zero(T)
+  tmax = typemax(T)
+
+  # check for intersection with slabs along with each axis
+  for i in 1:Dim
+    imin = (lo[i] - orig[i]) * invdir[i]
+    imax = (up[i] - orig[i]) * invdir[i]
+
+    # swap variables if necessary
+    invdir[i] < zero(T) && ((imin, imax) = (imax, imin))
+
+    # the ray is on a face of the box, avoid NaN
+    (isnan(imin) || isnan(imax)) && continue
+
+    (tmin > imax || imin > tmax) && return NoIntersection() |> f
+    
+    tmin = max(tmin, imin)
+    tmax = min(tmax, imax)
+  end
+
+  if tmin ≈ tmax
+    return TouchingRayBox(r(tmin)) |> f
+  elseif tmin < tmax
+    return CrossingRayBox(Segment(r(tmin), r(tmax))) |> f
+  end
```

Comment:
Can you please double check the return type of the function with `@code_warntype`? Now that we don't have a final `else` clause, Julia may be returning a `Union{Intersection,Nothing}` type. We should probably add a dummy `else` clause with `NoIntersection` as before even after the special intersection fix.

Reply:
@juliohm It is impossible to mis-hit on the elseif because we checked NaN in previous code. Now, I changed the code format to avoid this elseif.

Reply:
And the result of `@code_warntype` is ok? I think we can merge after that 👍🏽

Reply:
@juliohm 

1 red line:

```julia
Body::Union{NoIntersection, CrossingRayBox{Segment{3, Float64, StaticArrays.SVector{2, Point3}}}, TouchingRayBox{Point3}}
```

1 yellow line:

```julia
@_5::Union{Nothing, Tuple{Int64, Int64}}
```

full code:

```julia
Variables
  #self#::Core.Const(Meshes.intersecttype)
  f::Core.Const(identity)
  r::Ray{3, Float64}
  b::Box{3, Float64}
  @_5::Union{Nothing, Tuple{Int64, Int64}}
  @_6::Int64
  tmax::Float64
  tmin::Float64
  orig::StaticArrays.SVector{3, Float64}
  up::StaticArrays.SVector{3, Float64}
  lo::StaticArrays.SVector{3, Float64}
  invdir::StaticArrays.SVector{3, Float64}
  i::Int64
  imax::Float64
  imin::Float64

Body::Union{NoIntersection, CrossingRayBox{Segment{3, Float64, StaticArrays.SVector{2, Point3}}}, TouchingRayBox{Point3}}
1 ── %1  = Meshes.one($(Expr(:static_parameter, 2)))::Core.Const(1.0)
│    %2  = Meshes.direction(r)::StaticArrays.SVector{3, Float64}
│    %3  = Base.broadcasted(Meshes.:/, %1, %2)::Core.PartialStruct(Base.Broadcast.Broadcasted{StaticArrays.StaticArrayStyle{1}, Nothing, typeof(/), Tuple{Float64, StaticArrays.SVector{3, Float64}}}, Any[Core.Const(/), Core.PartialStruct(Tuple{Float64, StaticArrays.SVector{3, Float64}}, Any[Core.Const(1.0), StaticArrays.SVector{3, Float64}]), Core.Const(nothing)])
│          (invdir = Base.materialize(%3))
│    %5  = Meshes.extrema(b)::Tuple{Point3, Point3}
│    %6  = Base.broadcasted(Meshes.coordinates, %5)::Base.Broadcast.Broadcasted{Base.Broadcast.Style{Tuple}, Nothing, typeof(coordinates), Tuple{Tuple{Point3, Point3}}}
│    %7  = Base.materialize(%6)::Tuple{StaticArrays.SVector{3, Float64}, StaticArrays.SVector{3, Float64}}
│    %8  = Base.indexed_iterate(%7, 1)::Core.PartialStruct(Tuple{StaticArrays.SVector{3, Float64}, Int64}, Any[StaticArrays.SVector{3, Float64}, Core.Const(2)])
│          (lo = Core.getfield(%8, 1))
│          (@_6 = Core.getfield(%8, 2))
│    %11 = Base.indexed_iterate(%7, 2, @_6::Core.Const(2))::Core.PartialStruct(Tuple{StaticArrays.SVector{3, Float64}, Int64}, Any[StaticArrays.SVector{3, Float64}, Core.Const(3)])
│          (up = Core.getfield(%11, 1))
│    %13 = Meshes.origin(r)::Point3
│          (orig = Meshes.coordinates(%13))
│          (tmin = Meshes.zero($(Expr(:static_parameter, 2))))
│          (tmax = Meshes.typemax($(Expr(:static_parameter, 2))))
│    %17 = (1:$(Expr(:static_parameter, 1)))::Core.Const(1:3)
│          (@_5 = Base.iterate(%17))
│    %19 = (@_5::Core.Const((1, 1)) === nothing)::Core.Const(false)
│    %20 = Base.not_int(%19)::Core.Const(true)
└───       goto #15 if not %20
2 ┄─ %22 = @_5::Tuple{Int64, Int64}::Tuple{Int64, Int64}
│          (i = Core.getfield(%22, 1))
│    %24 = Core.getfield(%22, 2)::Int64
│    %25 = Base.getindex(lo, i)::Float64
│    %26 = Base.getindex(orig, i)::Float64
│    %27 = (%25 - %26)::Float64
│    %28 = Base.getindex(invdir, i)::Float64
│          (imin = %27 * %28)
│    %30 = Base.getindex(up, i)::Float64
│    %31 = Base.getindex(orig, i)::Float64
│    %32 = (%30 - %31)::Float64
│    %33 = Base.getindex(invdir, i)::Float64
│          (imax = %32 * %33)
│    %35 = Base.getindex(invdir, i)::Float64
│    %36 = Meshes.zero($(Expr(:static_parameter, 2)))::Core.Const(0.0)
│    %37 = (%35 < %36)::Bool
└───       goto #4 if not %37
3 ── %39 = imax::Float64
│    %40 = imin::Float64
│          (imin = %39)
│          (imax = %40)
└───       goto #4
4 ┄─ %44 = Meshes.isnan(imin)::Bool
└───       goto #6 if not %44
5 ──       goto #7
6 ── %47 = Meshes.isnan(imax)::Bool
└───       goto #8 if not %47
7 ┄─       goto #13
8 ── %50 = (tmin > imax)::Bool
└───       goto #10 if not %50
9 ──       goto #11
10 ─ %53 = (imin > tmax)::Bool
└───       goto #12 if not %53
11 ┄ %55 = Meshes.NoIntersection()::Core.Const(NoIntersection())
│    %56 = (%55 |> f)::Core.Const(NoIntersection())
└───       return %56
12 ─       (tmin = Meshes.max(tmin, imin))
└───       (tmax = Meshes.min(tmax, imax))
13 ┄       (@_5 = Base.iterate(%17, %24))
│    %61 = (@_5 === nothing)::Bool
│    %62 = Base.not_int(%61)::Bool
└───       goto #15 if not %62
14 ─       goto #2
15 ┄ %65 = (tmin ≈ tmax)::Bool
└───       goto #17 if not %65
16 ─ %67 = (r)(tmin)::Point3
│    %68 = Meshes.TouchingRayBox(%67)::TouchingRayBox{Point3}
│    %69 = (%68 |> f)::TouchingRayBox{Point3}
└───       return %69
17 ─ %71 = (r)(tmin)::Point3
│    %72 = (r)(tmax)::Point3
│    %73 = Meshes.Segment(%71, %72)::Segment{3, Float64, StaticArrays.SVector{2, Point3}}
│    %74 = Meshes.CrossingRayBox(%73)::CrossingRayBox{Segment{3, Float64, StaticArrays.SVector{2, Point3}}}
│    %75 = (%74 |> f)::CrossingRayBox{Segment{3, Float64, StaticArrays.SVector{2, Point3}}}
└───       return %75
```

Reply:
I think it is ok, it is returning an union of intersection types. We will fix that in the next release by refactoring the internal intersection API.